### ISSUE TITLE
DataSourcePlugin: Need a migration handler for datasource plugins

### ISSUE THREAD

**Issue by srclosson** [grafana/grafana#31318]
**What would you like to be added**:
Panel plugins have a `setMigrationHandler` callback that can be set in `module.ts` that allows migrations to be run to the panels config on startup. It would be nice to have something similar for a datasource plugin.

**Why is this needed**:
As vendors api's mature and new features are added, datasources will likely evolve, adding fields to the query definition. To ensure backwards compatibility, being able to migrate a previous datasource query to the new "enhanced version" and have those changes persist would be valuable. 

There are at least 3 enterprise plugins I can think of now that would benefit from this feature. 

Other things to note:
* If a default cannot be assumed, it may be required to ask the user what they want. (so far, I cannot actually think of a case where we were not able to assume the default with an enterprise plugin, but you never know)

**Comment by sunker** [comment]
related to #30629

**Comment by sunker** [comment]
> There are at least 3 enterprise plugins I can think of now that would benefit from this feature.

I think it's become a common anti-pattern in data source plugins that migrations are spread out all over the data source frontend code. They can be found in datasource.query function and in the editors by manipulating the props. So if we had proper migration support it would be a major improvement for data source plugin developers.

**Comment by torkelo** [comment]
Question is where this migration handler should live (think that one is easy, DataSourceClass). When should it be called and from where? 

* When queries are added (to set defaults) 
* ???

**Comment by kminehart** [comment]
Why wouldn't it be called when the plugin is updated? Is that possible?

**Comment by joshhunt** [comment]
I guess only backend plugins would be able to benefit from this, but perhaps the backend could/should trigger the migration on Grafana startup/plugin init? This would ensure they're migrated for Alerts before the frontend is loaded by a user also.

**Comment by joshhunt** [comment]
> I think it's become a common anti-pattern in data source plugins that migrations are spread out all over the data source frontend code. They can be found in datasource.query function and in the editors by manipulating the props. 

Also @sunker I think there's also another slightly related concern of setting default values for queries as well. Sometimes a migration can be avoided if you

**Comment by kminehart** [comment]
I think that makes sense. My preferred method would be to handle it how SQL migrations are typically handled, only with Go or JavaScript instead of SQL.

```
migrations/
migrations/v7.5.0_migration.up.go
migrations/v7.5.0_migration.down.go
migrations/v7.5.1_migration.up.go
migrations/v7.5.1_migration.down.go
```

or something similar.

**Comment by Elfo404** [comment]
the Elasticsearch plugin would benefit a lot from this as well.

/cc. @simianhacker

**Comment by sunker** [comment]
> I guess only backend plugins would be able to benefit from this, but perhaps the backend could/should trigger the migration on Grafana startup/plugin init? This would ensure they're migrated for Alerts before the frontend is loaded by a user also.

Yeah I agree. But would be great if migrations were available for frontend plugins too. Also, I think the threshold for writing migrations would be much lower if they could be done in javascript instead of go.

**Comment by ryantxu** [comment]
Assuming things go well with the cue model definitions... that should be a single standard way to do migrations for panels and datasources.  `<handwave><handwave><handwave>`

**Comment by srclosson** [comment]
A few thoughts:

I like the idea of having a `setMigrationHandler`. I thought there were some good points in this article:

https://levelup.gitconnected.com/an-approach-to-javascript-object-schema-migration-ae1bd9f0ce78

So if the plugin version is different, calling the migration handler with the old schema version and the new schema version makes sense to me. Whatever schema is returned is saved as the dashboard JSON. Perhaps the user may want to be notified of the changes before committing them?

**Comment by torkelo** [comment]
not sure how this could be done in the backend via some schema migration that updates dashboard, it needs to be part of the query execution flow (so before executing). To do it on persisted dashboards would require updating all saved dashboards which would be very slow and expensive and would need scheduled downtime to complete (not acceptable for Grafana Cloud).

And for provisioned dashboards, when would this migration happen? if it happened before save it could cause a diff between saved dashboard and json file and cause endless updates. 

So it's pretty tricky. We also need something to get a default query model in the UI when creating new queries.

**Comment by tolzhabayev** [comment]
@srclosson there is a PR that might be addressing this issue https://github.com/grafana/grafana/pull/43901 - please check if it does.

**Comment by mckn** [comment]
I'm reopening this issue since this is something we need/want to solve long term.

**Comment by tolzhabayev** [comment]
https://grafana.com/developers/plugin-tools/how-to-guides/data-source-plugins/add-migration-handler-for-backend-data-source describes how to add query migration handlers to datasources

**Comment by mckn** [comment]
https://github.com/grafana/grafana/issues/31318#issuecomment-2866115965
